File: src/receipts/report.py
"""The evidence bundle: what the agent did, what it claimed, and where they differ."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from . import __version__
from .actions import Action, ActionKind
from .detectors.base import Finding, Severity
from .model import Trace

CLEAN = "clean"
DIVERGED = "diverged"


@dataclass(frozen=True, slots=True)
class Report:
    trace: Trace
    actions: tuple[Action, ...]
    findings: tuple[Finding, ...]

    @property
    def verdict(self) -> str:
        return DIVERGED if self.findings else CLEAN

    def worst(self) -> Severity | None:
        return min((f.severity for f in self.findings), key=lambda s: s.rank, default=None)

    def counts(self) -> dict[str, int]:
        tally = {str(s): 0 for s in Severity}
        for finding in self.findings:
            tally[str(finding.severity)] += 1
        return tally

    def as_dict(self) -> dict[str, Any]:
        stats = self.trace.stats
        return {
            "receipts_version": __version__,
            "verdict": self.verdict,
            "agent": self.trace.source,
            "run": {
                "status": self.trace.status,
                "task_id": stats.get("task_id"),
                "tool_calls": len(self.actions),
                "tokens": stats.get("total_tokens"),
                "cost": stats.get("session_costs"),
                "duration_ms": stats.get("duration_ms"),
            },
            "claim": self.trace.final_message,
            "counts": self.counts(),
            "findings": [f.as_dict() for f in self.findings],
            "ground_truth": {
                "commands": [
                    {"seq": a.seq, "command": a.target, "outcome": str(a.outcome)}
                    for a in self.actions
                    if a.kind is ActionKind.RUN_COMMAND
                ],
                "files_written": [
                    {"seq": a.seq, "path": path, "tool": a.tool_name}
                    for a in self.actions
                    for path in a.writes
                ],
            },
        }


def build(trace: Trace, actions: list[Action], findings: list[Finding]) -> Report:
    return Report(trace=trace, actions=tuple(actions), findings=tuple(findings))

File: site/lib/report.ts
import data from "./report.json";

export type Evidence = { seq: number; label: string; excerpt: string };
export type Finding = {
  detector: string;
  severity: string;
  title: string;
  detail: string;
  evidence: Evidence[];
};
export type TimelineRow = {
  seq: number;
  kind: string;
  detail: string;
  outcome: string;
  flagged: boolean;
  test: boolean;
  recovered: boolean;
};
export type Run = {
  name: string;
  scenario: string;
  label: string | null;
  verdict: string;
  agent: string;
  claim: string;
  findings: Finding[];
  files_written: number;
  commands: number;
  duration_ms: number | null;
  cost: number | null;
  timeline: TimelineRow[];
  strip: string[];
  traceLines: number;
  citedLines: number;
};

export type Report = {
  generated: string;
  receipts_version: string;
  agents: string[];
  scenarios: string[];
  totals: {
    diverged: number;
    clean: number;
    findings: number;
    trace_lines: number;
    cited_lines: number;
    false_alarms: number | null;
  };
  runs: Run[];
};

export const report = data as unknown as Report;

export const diverged = (r: Run) => r.verdict === "diverged";

export const runsByScenario = () => {
  const map = new Map<string, Run[]>();
  for (const run of report.runs) {
    const list = map.get(run.scenario) ?? [];
    list.push(run);
    map.set(run.scenario, list);
  }
  return map;
};
